activesupport/lib/active_support/cache/redis_cache_store.rb

```diff
@@ -0,0 +1,389 @@
+              unless entry.nil? || entry.expired? || entry.mismatched?(normalize_version(key, options))
+                results[name] = entry.value
+              end
+            end
+          end
+        end
+
+        # For compat with the wexpected options hash argument.
+        def write_entry(key, entry, options = nil)
+          write_entry_kwargs key, entry, **options
+        end
+
+        # Write an entry to the cache.
+        #
+        # Requires Redis 2.6.12+ for extended SET options.
+        def write_entry_kwargs(key, entry, unless_exist: false, raw: false, expires_in: nil, race_condition_ttl: nil, **options)
+          value = raw ? entry.value.to_s : serialize_entry(entry)
+
+          # If race condition TTL is in use, ensure that cache entries
+          # stick around a bit longer after they would have expired
+          # so we can purposefully serve stale entries.
+          if race_condition_ttl && expires_in && expires_in > 0 && !raw
+            expires_in += 5.minutes
+          end
+
+          modifiers = {}
+          modifiers[:nx] = unless_exist
+          modifiers[:px] = (1000 * expires_in.to_f).ceil if expires_in
+
+          failsafe :write_entry do
+            redis.set key, value, modifiers
+          end
+        end
+
+        # Delete an entry from the cache.
+        def delete_entry(key, options)
+          failsafe :delete_entry, returning: false do
+            redis.del key
+          end
+        end
+
+        # Nonstandard store provider API to write multiple values at once.
+        def write_multi_entries(entries, options)
+          if @mset_capable && options.none?
+            begin
+              write_multi_entries_mset(entries, options)
+            rescue Redis::Distributed::CannotDistribute
+              @mset_capable = false
+              super
+            end
+          else
+            super
+          end
+        end
+
+        def write_multi_entries_mset(entries)
+          if entries.any?
+            failsafe :write_entries do
+              redis.mapped_mset(entries)
+            end
+          end
+        end
+
+        # Truncate keys that exceed 1kB.
+        def normalize_key(key, options)
+          truncate_key super
+        end
+
+        def truncate_key(key)
+          if key.bytesize > max_key_bytesize
+            suffix = ":md5:#{Digest::MD5.hexdigest(key)}"
+            truncate_at = max_key_bytesize - suffix.bytesize
+            "#{key.byteslice(0, truncate_at)}#{suffix}"
+          else
+            key
+          end
+        end
+
+        def deserialize_entry(raw_value)
+          if raw_value
+            entry = Marshal.load(raw_value) rescue raw_value
+            entry.is_a?(Entry) ? entry : Entry.new(entry)
+          end
+        end
+
+        def serialize_entry(entry)
+          Marshal.dump(entry)
+        end
+
+        def failsafe(method, returning: nil)
+          yield
+        rescue ::Redis::BaseConnectionError => e
```

Comment:
May need to expand the set of failsafe exceptions, and allow Rails users to do it.